File: 0912-sort-an-array/0912-sort-an-array.cpp
class Solution {
public:
    void merge(vector<int>& nums, int start, int mid, int end) {
        int p = start, q = mid + 1, k = 0;
        vector<int> temp(end - start + 1);
        while(p <= mid && q <= end) {
            if (nums[p] < nums[q]) {
                temp[k++] = nums[p++];
            } else {
                temp[k++] = nums[q++];
            }
        }
        while(p <= mid){
            temp[k++] = nums[p++];
        }
        while(q <= end){
            temp[k++] = nums[q++];
        }
        for (int i = 0; i < k; i++) {
            nums[start++] = temp[i];
        }
    }
    void mergeSort(vector<int>& nums, int start, int end) {
        if (start < end) {
            int mid = (end - start) / 2 + start;
            mergeSort(nums, start, mid);
            mergeSort(nums, mid + 1, end);
            merge(nums, start, mid, end);
        }
    }
    vector<int> sortArray(vector<int>& nums) {
        mergeSort(nums, 0, nums.size() - 1);
        return nums;
    }
};

File: 0912-sort-an-array/0912-sort-an-array.py
class Solution:
    def mergeSort(self, nums, l, r):
        if l < r:
            m = (l + r) // 2
            self.mergeSort(nums, l, m)
            self.mergeSort(nums, m + 1, r)
            self.merge(nums, l, m, r)
    def merge(self, nums, l, m, r):
        i = l
        j = m + 1
        res = []
        while i <= m and j <= r:
            if nums[i] <= nums[j]:
                res.append(nums[i])
                i += 1
            else:
                res.append(nums[j])
                j += 1
        while i <= m:
            res.append(nums[i])
            i += 1
        while j <= r:
            res.append(nums[j])
            j += 1
        index = 0
        for k in range(l, r + 1):
            nums[k] = res[index]
            index += 1
        
    def sortArray(self, nums: List[int]) -> List[int]:
        self.mergeSort(nums, 0, len(nums) - 1)
        return nums